### PR TITLE
feat: plugin indexes

### PR DESCRIPTION
**Pull request by regisb**
We implement this TEP: https://discuss.openedx.org/t/tutor-enhancement-proposal-tep-plugin-indices/8182

With plugin indexes, tutor users can install and upgrade plugins directly from indexes:

    tutor plugins install ecommerce
    tutor plugins index add contrib
    tutor plugins install codejail
    tutor plugins upgrade all

This change has been long in the coming \o/

### CODE REVIEW

**Comment by regisb** [comment]
@overhangio/tutor-maintainers This PR is ready for a review. I think that the new plugin indexes that we introduce here are very important for plugin maintainers, so I would really like to have 3rd party opinions. For instance, do folks from eduNEXT or Opencraft see the point of plugin indexes? This is what they would look like, in the current implementation;

https://github.com/overhangio/tutor-plugins-index/blob/master/main/olive/plugins.yml
https://github.com/overhangio/tutor-plugins-index/blob/master/contrib/olive/plugins.yml

Would you consider building your own index? Would you like to add your plugins to the contrib index?

File: docs/plugins/intro.rst
Comment (arbrandes) [inline docs/plugins/intro.rst:43]:
```suggestion
Many plugins are available from plugin indexes. These indexes are lists of plugins, similar to the `pypi <https://pypi.org>`__ or `npm <npmjs.com/>`__ indexes. By default, Tutor comes with the "main" plugin index. You can check available plugins from this index by running::
```

File: docs/plugins/intro.rst
Comment (arbrandes) [inline docs/plugins/intro.rst:48]: When reading this the first time (before reading the code), it wasn't clear to me that "contrib" and "wizard" are keywords that Tutor will match against indexes defined in [tutor-plugins-index](https://github.com/overhangio/tutor-plugins-index).  Maybe we could add something along the lines of:

```suggestion
More plugins can be downloaded from the "contrib" and "wizard" indexes (both of which are defined in the `official index repository <https://github.com/overhangio/tutor-plugins-index>`__, and thus automatically recognized as such by Tutor)::
```

File: tutor/plugins/indexes.py
Comment (arbrandes) [inline tutor/plugins/indexes.py:119]: What happens if somebody removes one of the official indices?

**Review by arbrandes — COMMENTED** [review]
Did not actually test this yet, but made a couple of comments.  Looks great!

File: tutor/plugins/indexes.py
Comment (regisb) [reply tutor/plugins/indexes.py:119]: It's just like removing an apt or pypi index: nothing much happens. Installed plugins are not uninstalled or deactivated. But plugins will no longer be automatically upgraded with `tutor plugins upgrade all` and cannot be installed or uninstalled.

File: docs/plugins/intro.rst
Comment (regisb) [reply docs/plugins/intro.rst:48]: Good catch, I'll change the wording.

File: tutor/plugins/indexes.py
Comment (arbrandes) [reply tutor/plugins/indexes.py:119]: Makes sense, thanks!

**Comment by felipemontoya** [comment]
I like the idea of the plugin indexes quite a lot. I can't comment on the specifics of how it's implemented due to my lack of familiarity with the tutor-core code.

As a third party with a team I would probably build my own index so that I can share plugins easily with them. I also like the idea of the contrib index so that things that are used by the community broadly can found even more easily. I suppose that contrib index would be also installed by default but listing something there will require some form of vouching or vetting.

**Comment by regisb** [comment]
Thanks for commenting @felipemontoya! Do you agree with the addition of codejail to the contrib index? https://github.com/overhangio/tpi/blob/master/contrib/olive/plugins.yml Are there any other plugins you would like to add?

**Comment by Alec4r** [comment]
@regisb wow thanks, I have been waiting for this, but I've got a question for you, what happen with private plugins? will we able to add our own index?.

I mean, I would like add something like:

tutor plugins index add edunext https://GitHub.com/eduNEXT/tutor-indices.txt

Something like a git remote.

**Comment by regisb** [comment]
> what happen with private plugins? will we able to add our own index?

Yes :) First of all, the user should have access to the plugins.yml file. It can be accessible behind a VPN, a basic auth http://user:password@edunext.com/index/ url, or because they downloaded the index it to their disk (`tutor plugins index add` supports local paths).

Second, "src" parameter for each all plugins that are listed in the index are actually pip-compatible specs that will be read from a requirements.txt file. This means that you can add there any reference that follows this spec: https://pip.pypa.io/en/stable/reference/requirements-file-format/ In particular, you can write:

    # Python package from a private index
    src: |
        --index-url=https://pip.mymirror.org
        my-plugin>=10.0 

All of this is explained with better words in the docs: https://github.com/overhangio/tutor/pull/783/files#diff-d711f720e3c5a7d95c0d3a3bfce00b6aa51a4377c94306328c95893693586de1

**Comment by felipemontoya** [comment]
I definitely agree that codejail should be in contrib. Ideally lots of users for that one.

I think that codejail is the only plugin maintained by edunext that should land there at the moment. I hope we will be making more in the future that are broad like this, but we have to make them first ;)

**Comment by regisb** [comment]
@kdmccormick I'd like to merge this -- unless you want to review it first? Let me know.

**Comment by kdmccormick** [comment]
@regisb Yes, I would like to do a quick review, I will do it within the next two hours. Thanks for the reminder.

File: tutor/commands/plugins.py
Comment (kdmccormick) [inline tutor/commands/plugins.py:362]:
```suggestion
@click.command(name="list", help="List plugin indexes")
```
Both words are correct, but you're using "indexes" everywhere else so I recommend sticking to that.

File: tutor/commands/plugins.py
Comment (kdmccormick) [inline tutor/commands/plugins.py:328]:
```suggestion
    raise exceptions.TutorError(f"Plugin could not be found in indexes: {name}")
```

If I run this on a locally-installed plugin that's not in an index, then I will see this error. I think it'd be clearer to be told that the plugin is "not found in an index" than "not found".

**Review by kdmccormick — APPROVED** [review]
Fantastic!

File: tutor/commands/plugins.py
Comment (regisb) [reply tutor/commands/plugins.py:362]: Good catch! I changed the terminology after I wrote the TEP.

**Comment by regisb** [comment]
Thanks for the quick review Kyle! I'll merge this now.

**Pull request merged**
